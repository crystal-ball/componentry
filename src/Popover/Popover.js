import activeContainer from '../active-container-factory'
import activeContent from '../active-content-factory'
import activeTrigger from '../active-trigger-factory'
import elem from '../elem-factory'
import { useTheme } from '../Theme/Theme'

const Popover = activeContainer('popover', {
  direction: 'right',
  escEvents: true,
  mouseEvents: true,
})
export default Popover

Popover.Body = function PopoverBody(props) {
  return elem({
    componentClassNames: 'popover-body',
    ...useTheme('PopoverBody'),
    ...props,
  })
}
Popover.Body.displayName = 'PopoverBody'

Popover.Content = activeContent('popover', {
  arias: { id: true, role: 'tooltip', hidden: true },
  popper: true,
})

Popover.Heading = function PopoverHeading(props) {
  return elem({
    as: 'h3',
    componentClassNames: 'popover-heading',
    ...useTheme('PopoverHeading'),
    ...props,
  })
}
Popover.Heading.displayName = 'PopoverHeading'

Popover.Trigger = activeTrigger('popover', {
  arias: { describedby: true },
})
