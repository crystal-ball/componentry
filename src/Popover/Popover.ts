import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionBuilder } from '../utils/active-action-component-builder'
import { activeContentBuilder } from '../utils/active-content-component-builder'
import { staticComponent } from '../utils/static-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
  ComponentBaseProps,
} from '../utils/types'

export interface PopoverProps
  extends ActiveContainerBaseProps,
    ComponentBaseProps<'div'> {}

export interface PopoverActionProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverBodyProps
  extends ActiveContentBaseProps,
    ComponentBaseProps<'div'> {}

export interface PopoverContentProps extends ComponentBaseProps<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverHeadingProps extends ComponentBaseProps<'div'> {}

export interface Popover {
  (props: PopoverProps): React.ReactElement
  /**
   * [Popover action component üìù](https://componentry.design/components/popover)
   */
  Action: React.FC<PopoverActionProps>
  /**
   * [Popover body component üìù](https://componentry.design/components/popover)
   */
  Body: React.FC<PopoverBodyProps>
  /**
   * [Popover content component üìù](https://componentry.design/components/popover)
   */
  Content: React.FC<PopoverContentProps>
  /**
   * [Popover heading component üìù](https://componentry.design/components/popover)
   */
  Heading: React.FC<PopoverHeadingProps>
}

/**
 * [Popover component üìù](https://componentry.design/components/popover)
 * @experimental
 */
export const Popover = activeContainerBuilder<PopoverProps>('Popover', {
  direction: 'right',
  escEvents: true,
  mouseEvents: true,
}) as Popover

Popover.Action = activeActionBuilder<PopoverActionProps>('PopoverAction', {
  aria: { describedby: true },
})

Popover.Body = staticComponent<PopoverBodyProps>('PopoverBody')

Popover.Content = activeContentBuilder<PopoverContentProps>('PopoverContent', {
  aria: { id: true, role: 'tooltip', hidden: true },
  positioned: true,
})

Popover.Heading = staticComponent<PopoverHeadingProps>('PopoverHeading', {
  as: 'h3',
})
