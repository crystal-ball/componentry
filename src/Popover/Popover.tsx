import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionComponent } from '../utils/active-action-component-builder'
import { activeContentComponent } from '../utils/active-content-component-builder'
import { staticComponent } from '../utils/static-component-builder'
import {
  BaseActiveActionProps,
  BaseActiveContainerProps,
  BaseActiveContentProps,
  BaseProps,
} from '../utils/base-types'

export interface PopoverProps
  extends BaseActiveContainerProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface PopoverActionProps
  extends BaseActiveActionProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'button'>, 'className'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverBodyProps
  extends BaseActiveContentProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface PopoverContentProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverHeadingProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface Popover {
  (props: PopoverProps): React.ReactElement
  /**
   * [Popover action component üìù](https://componentry.design/components/popover)
   */
  Action: React.FC<PopoverActionProps>
  /**
   * [Popover body component üìù](https://componentry.design/components/popover)
   */
  Body: React.FC<PopoverBodyProps>
  /**
   * [Popover content component üìù](https://componentry.design/components/popover)
   */
  Content: React.FC<PopoverContentProps>
  /**
   * [Popover heading component üìù](https://componentry.design/components/popover)
   */
  Heading: React.FC<PopoverHeadingProps>
}

/**
 * [Popover component üìù](https://componentry.design/components/popover)
 */
export const Popover = activeContainerBuilder<PopoverProps>('Popover', {
  direction: 'right',
  escEvents: true,
  mouseEvents: true,
}) as Popover

Popover.Action = activeActionComponent('PopoverAction', {
  aria: { describedby: true },
})

Popover.Body = staticComponent<PopoverBodyProps>('PopoverBody', {
  componentCx: 'üÖ≤-popover-body',
})

Popover.Content = activeContentComponent<PopoverContentProps>('PopoverContent', {
  aria: { id: true, role: 'tooltip', hidden: true },
  positioned: true,
})

Popover.Heading = staticComponent<PopoverHeadingProps>('PopoverHeading', {
  as: 'h3',
  componentCx: 'üÖ≤-popover-heading',
})
