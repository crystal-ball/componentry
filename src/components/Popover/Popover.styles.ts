import { getMergedConfig } from '../../plugin-postcss/configs'

const { theme } = getMergedConfig()

//                                       <Popover /> styles
// --------------------------------------------------------

const arrowWidth = 4 // in pixels

export const popoverStyles = {
  // --- CONTAINER
  '.C9Y-Popover-base': {
    // Default align content to dead center of action the specific content styles can
    // then easily move to top/left/right/bottom where needed
    alignItems: 'center',
    display: 'inline-flex', // flex instead of block for easier center alignment of content
    justifyContent: 'center',
    position: 'relative',

    // Directional styles

    '&.C9Y-Popover-top': {
      '.C9Y-PopoverContent': {
        bottom: '100%',
        justifyContent: 'center',
      },

      '.C9Y-PopoverContentContents': {
        // Top requires bumping bottom 100% again to compensate for container not having
        // any height
        bottom: '100%',
        marginBottom: `${arrowWidth}px`,
      },

      '.C9Y-PopoverArrow:after': {
        // top: 100%,
        // width: 100%,
        // }
        bottom: `${arrowWidth * 0.5}px`,
      },
    },

    '&.C9Y-Popover-right': {
      '.C9Y-PopoverContent': {
        left: '100%',
        alignItems: 'center',
      },

      '.C9Y-PopoverContentContents': {
        marginLeft: `${arrowWidth}px`,
      },

      // '.C9Y-PopoverArrow': {
      //   height: '100%',
      //   right: '100%',
      // },

      '.C9Y-PopoverArrow:after': {
        left: `${arrowWidth * 0.5}px`,
      },
    },

    '&.C9Y-Popover-bottom': {
      '.C9Y-PopoverContent': {
        top: '100%',
        justifyContent: 'center',
      },

      '.C9Y-PopoverContentContents': {
        marginTop: `${arrowWidth}px`,
      },

      // '.tipContainer': {
      //   bottom: '100%',
      //   width: '100%',
      // },

      '.C9Y-PopoverArrow:after': {
        top: `${arrowWidth * 0.5}px`,
      },
    },

    '&.C9Y-Popover-left': {
      '.C9Y-PopoverContent': {
        right: '100%',
        alignItems: 'center',
      },

      '.C9Y-PopoverContentContents': {
        marginRight: `${arrowWidth}px`,
        right: 0,
      },

      // '.tip-container': {
      //   height: '100%',
      //   left: '100%',
      // },

      '.C9Y-PopoverArrow:after': {
        right: `${arrowWidth * 0.5}px`,
      },
    },
  },

  // --- Popover action ------------------------------------

  '.C9Y-PopoverAction': {},

  // Content container overrides the width constraints for parent element
  '.C9Y-PopoverContent': {
    width: '300px',
    position: 'absolute',
    display: 'inline-flex',
  },

  // Content Element - The content element should be used to wrap a body/header element
  '.C9Y-PopoverContentContents': {
    display: 'block',
    position: 'absolute',
    maxWidth: '300px',
    backgroundClip: 'padding-box',
    backgroundColor: theme.colors.background,
    border: theme.border.DEFAULT,
    fontSize: theme.fontSize.sm,
    // Allow breaking very long words so they don't overflow the popover's bounds
    wordWrap: 'break-word',

    // box-shadow: box-shadow($popover-box-shadow),
    borderRadius: theme.borderRadius.DEFAULT,

    // '.tip-container': {
    //   alignItems: 'center',
    //   display: 'flex',
    //   justifyContent: 'center',
    //   overflow: 'hidden',
    //   position: 'absolute',
    // },
  },

  // The .tip container creates a containing box of the same height and width as the
  // tip with overflow hidden. Then inside that container we create a square of the
  // same dimensions, and rotate it 45 degrees to create the tip effect
  '.C9Y-PopoverArrow': {
    height: `${arrowWidth}px`,
    overflow: 'hidden',
    width: `${arrowWidth}px`,

    '&:after': {
      border: theme.border.DEFAULT,
      // box-shadow: $popover-box-shadow,
      background: theme.colors.background,
      content: "''",
      height: `${arrowWidth}px`,
      position: 'absolute',
      transform: 'rotate(45deg)',
      width: `${arrowWidth}px`,
    },
  },

  // --- BODY
  // Body element should contain popover inner content
  '.C9Y-PopoverBody': {
    padding: theme.spacing[4],
    color: theme.colors.gray[700],
  },

  // --- HEADING
  '.C9Y-PopoverHeading': {
    padding: theme.spacing[2],
    // Try to guarantee that spacing is correct regardless of div vs h* usage
    marginTop: 0,
    marginBottom: 0,
    fontSize: theme.fontSize.lg,
    color: theme.colors.gray[900],
    borderBottom: theme.border.DEFAULT,
  },
}
