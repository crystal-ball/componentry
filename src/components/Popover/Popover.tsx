import React from 'react'
import { activeActionBuilder } from '../../utils/active-action-component-builder'
import { activeContainerBuilder } from '../../utils/active-container-component-builder'
import { activeContentBuilder } from '../../utils/active-content-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
} from '../../utils/base-types'
import { staticComponent } from '../../utils/static-component-builder'
import { UtilityProps } from '../../utils/utility-classes'
import { Button } from '../Button/Button'

export interface PopoverProps
  extends ActiveContainerBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {}

export interface PopoverActionProps
  extends ActiveActionBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverBodyProps
  extends ActiveContentBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {}

export interface PopoverContentProps
  extends UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface PopoverHeadingProps
  extends UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {
  as?: React.ElementType
}

export interface Popover {
  (props: PopoverProps): React.ReactElement
  /**
   * [Popover action component üìù](https://componentry.design/components/popover)
   */
  Action: React.FC<PopoverActionProps>
  /**
   * [Popover body component üìù](https://componentry.design/components/popover)
   */
  Body: React.FC<PopoverBodyProps>
  /**
   * [Popover content component üìù](https://componentry.design/components/popover)
   */
  Content: React.FC<PopoverContentProps>
  /**
   * [Popover heading component üìù](https://componentry.design/components/popover)
   */
  Heading: React.FC<PopoverHeadingProps>
}

/**
 * [Popover component üìù](https://componentry.design/components/popover)
 * @experimental
 */
export const Popover = activeContainerBuilder<PopoverProps>('Popover', {
  direction: 'right',
  escEvents: true,
  mouseEvents: true,
}) as Popover

Popover.Action = activeActionBuilder<PopoverActionProps>('PopoverAction', {
  aria: { describedby: true },
  defaultAs: Button,
})

function PopoverContentElement({
  children,
  // renderArrow = false, TODO
  ...rest
}: {
  children: React.ReactNode
  // renderArrow?: boolean
}) {
  return (
    <div {...rest}>
      {/* {renderArrow && <div className='C9Y-PopoverArrow' />} */}
      <div className='C9Y-PopoverContentContents'>{children}</div>
    </div>
  )
}

Popover.Content = activeContentBuilder<PopoverContentProps>('PopoverContent', {
  aria: { id: true, role: 'tooltip', hidden: true },
  defaultAs: PopoverContentElement,
})

Popover.Heading = staticComponent<PopoverHeadingProps>('PopoverHeading', {
  as: 'h3',
})

Popover.Body = staticComponent<PopoverBodyProps>('PopoverBody')
