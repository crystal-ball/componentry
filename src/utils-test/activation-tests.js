/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * This set of activation event tests should pass for every element with active
 * state. It's probably not really necessary to run these tests for each element,
 * but this provides a simple high level gut check on if functionality is working
 * as expected for each component.
 *
 * IMPORTANT NOTES
 * - `dt()` is a shorthand method for creating data-test strings. Using data-test
 *   strings lets us reuse shared tests across components
 * - Enzyme `find()` will find **EVERY** instance of a prop. Since we are passing
 *   data-test attributes down here, it will always find more than one. Use
 *   `.last()` to selecte final node.
 */

const findContent = wrapper => wrapper.find(dt('content')).last()
const findTrigger = wrapper => wrapper.find(dt('trigger')).last()

export default TestComponent => {
  /**
   * Test that by default Active.Trigger elements render as button nodes, and
   * when passed an href render anchor nodes.
   */
  test('should render button or anchor tags by default', () => {
    const wrapper = mount(
      <TestComponent>
        <TestComponent.Trigger data-test="btn" />
        <TestComponent.Trigger data-test="anchor" href="http://localhost:3000/">
          Anchor
        </TestComponent.Trigger>
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    const btn = wrapper.find(dt('btn')).last()
    const anchor = wrapper.find(dt('anchor')).last()

    expect(anchor.type()).toEqual('a')
    expect(anchor.prop('type')).toEqual(undefined) // should not have type="button"
    expect(btn.type()).toEqual('button')
    expect(btn.prop('type')).toEqual('button') // should have type="button"
  })

  /**
   * Test that the default uncontrolled state scenario works. The active state should
   * be managed internally if not overriden with props and clicking the trigger should
   * show/hide content.
   */
  test('should update arias when trigger is activated', () => {
    const wrapper = mount(
      <TestComponent>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    // Content
    const content = findContent(wrapper)
    const guid = content.prop('id')
    expect(content.length).toEqual(1)
    expect(content.prop('aria-hidden')).toEqual('true')

    // Trigger
    const trigger = findTrigger(wrapper)
    expect(trigger.length).toEqual(1)

    // TODO: need to figure out common aria attributes and test only those here
    // eg Popover and tooltip do not have aria-controls attrs
    if (trigger.prop('aria-controls')) expect(trigger.prop('aria-controls')).toEqual(guid)

    // Simulate a click, arias should update
    trigger.simulate('click')

    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')

    // Simulate a click, arias should update
    trigger.simulate('click')

    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')
  })

  /**
   * Test that all of the user hooks are being called on activation/deactivation.
   * TODO: update to check that event hooks are called with correct args.
   */
  test('should call user event hooks', () => {
    const onActivate = jest.fn()
    const onActivated = jest.fn()
    const onDeactivate = jest.fn()
    const onDeactivated = jest.fn()
    const wrapper = mount(
      <TestComponent
        onActivate={onActivate}
        onActivated={onActivated}
        onDeactivate={onDeactivate}
        onDeactivated={onDeactivated}
      >
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    const trigger = findTrigger(wrapper)

    trigger.simulate('click')

    expect(onActivate).toHaveBeenCalledTimes(1)
    expect(onActivated).toHaveBeenCalledTimes(1)
    expect(onDeactivate).toHaveBeenCalledTimes(0) // not yet!
    expect(onDeactivated).toHaveBeenCalledTimes(0) // not yet!

    trigger.simulate('click')

    expect(onActivate).toHaveBeenCalledTimes(1) // only once
    expect(onActivated).toHaveBeenCalledTimes(1) // only once
    expect(onDeactivate).toHaveBeenCalledTimes(1)
    expect(onDeactivated).toHaveBeenCalledTimes(1)
  })

  /**
   * Test setting initial active state using defaultActive prop
   */
  test('state should default to defaultActive when passed', () => {
    const wrapper = mount(
      <TestComponent defaultActive>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')
  })

  /**
   * State checks if active prop value is different from internal state and updates
   * if that's the case, but this should only happen when passed active is an actual
   * value and not undefined. Check this with an instance with no value, and a
   * forced rerender
   */
  test('should not use active prop when value is undefined', () => {
    const wrapper = mount(
      <TestComponent>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')
    wrapper.update()
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')
  })

  /**
   * Test controlled component for methods `activate` and `deactivate`. Passing these
   * should override internal activate/deactivate meaning that clicking the trigger
   * for an element should only call those passed props.
   */
  test('should use passed activate and deactivate functions', () => {
    const activate = jest.fn()
    const deactivate = jest.fn()
    const wrapper = mount(
      <TestComponent activate={activate} deactivate={deactivate}>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    const trigger = findTrigger(wrapper)

    trigger.simulate('click')

    // We have overridden the activate event so, dropdown should still be closed
    // and clicking trigger should call activate again

    expect(activate).toHaveBeenCalledTimes(1)
    expect(deactivate).toHaveBeenCalledTimes(0)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')

    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(0)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')

    // Mock controlled component being passed true, click should now ONLY call deactivate
    wrapper.setProps({ active: true })
    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(1)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')

    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(2)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')
  })

  /**
   * Test that any flavor of the `<Active />` component exposes internal state and
   * state change methods using the FaCC pattern
   */
  test('should expose internals using FaCC pattern', () => {
    const wrapper = mount(
      <TestComponent>
        {({ active, activate, deactivate }) => (
          <div>
            <span data-test="active">{String(active)}</span>
            <button type="button" onClick={activate} data-test="activate" />
            <button type="button" onClick={deactivate} data-test="deactivate" />
          </div>
        )}
      </TestComponent>,
    )

    expect(wrapper.find(dt('active')).text()).toEqual('false')
    wrapper.find(dt('activate')).simulate('click')
    expect(wrapper.find(dt('active')).text()).toEqual('true')
  })
}
