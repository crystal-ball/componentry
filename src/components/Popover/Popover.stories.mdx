import { Meta, Story } from '@storybook/addon-docs'

import { Popover } from './Popover'

<Meta title='Components/Popover' component={Popover} />

# Popover

Popovers are dialogs that may contain interactive elements.

<Story name='Popover'>
  <Popover active direction='bottom'>
    <Popover.Action>Toggle</Popover.Action>
    <Popover.Content>
      <Popover.Heading>Fun Fact!</Popover.Heading>
      <Popover.Body>
        The new Texas Instrument calculators have ABC keyboards because if they had QWERTY
        keyboards, they would be considered computers and wouldnâ€™t be allowed for
        standardized test taking.
      </Popover.Body>
    </Popover.Content>
  </Popover>
</Story>
