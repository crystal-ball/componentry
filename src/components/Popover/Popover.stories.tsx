import type { Meta, StoryObj } from '@storybook/react'

import { Popover } from './Popover'

const meta: Meta<typeof Popover> = {
  component: Popover,
}

export default meta
type Story = StoryObj<typeof Popover>

export const Primary: Story = {
  render: () => (
    <Popover active direction='bottom'>
      <Popover.Action>Toggle</Popover.Action>
      <Popover.Content>
        <Popover.Heading>Fun Fact!</Popover.Heading>
        <Popover.Body>
          The new Texas Instrument calculators have ABC keyboards because if they had
          QWERTY keyboards, they would be considered computers and wouldnâ€™t be allowed for
          standardized test taking.
        </Popover.Body>
      </Popover.Content>
    </Popover>
  ),
}
