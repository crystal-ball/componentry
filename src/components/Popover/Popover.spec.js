import React from 'react'
import { render, screen } from '@testing-library/react'

import { activationTests } from '../../test/activation-tests'
import { elementTests } from '../../test/element-tests'
import { Popover } from './Popover'

describe('<Popover />', () => {
  // Basic library activation test suite
  activationTests(Popover, { name: 'popover', testArias: ['describedby'] })
  // Basic library element test suite
  elementTests(Popover)
  elementTests(Popover.Action)
  elementTests(Popover.Content)
  elementTests(Popover.Heading)
  elementTests(Popover.Body)

  it('renders the correct directional classes using direction', () => {
    render(
      <Popover data-testid='popover' direction='left'>
        <Popover.Action>Action</Popover.Action>
        <Popover.Content>Content</Popover.Content>
      </Popover>,
    )

    expect(screen.getByTestId('popover')).toHaveClass('left')
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Popover /> snapshots', () => {
  it('renders correctly', () => {
    render(
      <Popover data-testid='popover'>
        <Popover.Action>Toggle</Popover.Action>
        <Popover.Content>
          <Popover.Heading>Fun Fact!</Popover.Heading>
          <Popover.Body>
            <span>
              The new Texas Instrument calculators have ABC keyboards because if they had
              QWERTY keyboards, they would be considered computers and wouldnâ€™t be allowed
              for standardized test taking.
            </span>
          </Popover.Body>
        </Popover.Content>
      </Popover>,
    )

    expect(screen.getByTestId('popover')).toMatchSnapshot()
  })
})
