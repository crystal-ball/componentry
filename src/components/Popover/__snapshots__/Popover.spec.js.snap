// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popover /> snapshots renders correctly 1`] = `
<div
  class="right"
  data-id="test-guid"
  data-testid="popover"
>
  <button
    aria-describedby="test-guid"
    class="popover-action-primary"
    type="button"
  >
    Toggle
  </button>
  <div
    class="popover-content-container"
  >
    <div
      aria-hidden="true"
      class="popover-content-primary"
      id="test-guid"
      role="tooltip"
    >
      <div
        class="tip-container"
      >
        <div
          class="tip"
        />
      </div>
      <h3
        class=""
      >
        Fun Fact!
      </h3>
      <div
        class=""
      >
        <span>
          The new Texas Instrument calculators have ABC keyboards because if they had QWERTY keyboards, they would be considered computers and wouldnâ€™t be allowed for standardized test taking.
        </span>
      </div>
    </div>
  </div>
</div>
`;
